source/event_stream.c

```diff
@@ -0,0 +1,1118 @@
+
+    if (running_crc != message_crc) {
+        return aws_raise_error(AWS_ERROR_EVENT_STREAM_MESSAGE_CHECKSUM_FAILURE);
+    }
+
+    message->message_buffer = buffer;
+
+    if (aws_event_stream_message_headers_len(message) >
+        message_length - AWS_EVENT_STREAM_PRELUDE_LENGTH - AWS_EVENT_STREAM_TRAILER_LENGTH) {
+        message->message_buffer = 0;
+        return aws_raise_error(AWS_ERROR_EVENT_STREAM_MESSAGE_INVALID_HEADERS_LEN);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+//Verify buffer crcs and that length fields are reasonable. Once that is done, the buffer is copied to the message.
+int aws_event_stream_message_from_buffer_copy(struct aws_event_stream_message *message,
+                                              struct aws_allocator *alloc,
+                                              const uint8_t *buffer, size_t buffer_len) {
+    int parse_value = aws_event_stream_message_from_buffer(message, alloc, (uint8_t *) buffer, buffer_len);
+
+    if (!parse_value) {
+        message->message_buffer = aws_mem_acquire(alloc, buffer_len);
+
+        if (message->message_buffer) {
+            memcpy(message->message_buffer, buffer, buffer_len);
+            message->alloc = alloc;
+            message->owns_buffer = 1;
+
+            return AWS_OP_SUCCESS;
+        }
+
+        return aws_raise_error(AWS_ERROR_OOM);
+    }
+
+    return parse_value;
+}
+
+// if buffer is owned, release the memory.
+void aws_event_stream_message_clean_up(struct aws_event_stream_message *message) {
+    if (message->message_buffer && message->owns_buffer) {
+        aws_mem_release(message->alloc, message->message_buffer);
+    }
+}
+
+uint32_t aws_event_stream_message_total_length(const struct aws_event_stream_message *message) {
+    return aws_read_u32(message->message_buffer + TOTAL_LEN_OFFSET);
+}
+
+uint32_t aws_event_stream_message_headers_len(const struct aws_event_stream_message *message) {
+    return aws_read_u32(message->message_buffer + HEADER_LEN_OFFSET);
+}
+
+uint32_t aws_event_stream_message_prelude_crc(const struct aws_event_stream_message *message) {
+    return aws_read_u32(message->message_buffer + PRELUDE_CRC_OFFSET);
+}
+
+int aws_event_stream_message_headers(const struct aws_event_stream_message *message, struct aws_array_list *list) {
+    return get_headers_from_buffer(list,
+                                   message->message_buffer + AWS_EVENT_STREAM_PRELUDE_LENGTH,
+                                   aws_event_stream_message_headers_len(message));
+}
+
+const uint8_t *aws_event_stream_message_payload(const struct aws_event_stream_message *message) {
+    return message->message_buffer +
+           AWS_EVENT_STREAM_PRELUDE_LENGTH + aws_event_stream_message_headers_len(message);
+}
+
+uint32_t aws_event_stream_message_payload_len(const struct aws_event_stream_message *message) {
+    return aws_event_stream_message_total_length(message) -
```

Comment:
probably good idea to add underflow check here.

Reply:
This is handled per the other checks. Will add them elsewhere though.